airframe-http-finagle/src/main/scala/wvlet/airframe/http/finagle/FinagleRetryFilter.scala

```diff
@@ -0,0 +1,95 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package wvlet.airframe.http.finagle
+
+import com.twitter.finagle.http.{Request, Response, Status}
+import com.twitter.finagle.util.DefaultTimer
+import com.twitter.finagle.{Service, SimpleFilter, http}
+import com.twitter.util._
+import wvlet.airframe.control.ResultClass
+import wvlet.airframe.control.Retry.RetryContext
+import wvlet.log.LogSupport
+
+/**
+  * A filter for integrating Airframe Retry and Finagle
+  */
+class FinagleRetryFilter(retry: RetryContext, timer: Timer = DefaultTimer)
+    extends SimpleFilter[http.Request, http.Response]
+    with LogSupport {
+  import com.twitter.conversions.DurationOps._
+
+  private[this] def schedule(d: Duration)(f: => Future[Response]) = {
+    if (d > 0.seconds) {
+      val promise = new Promise[Response]
+      timer.schedule(Time.now + d) {
+        promise.become(f)
+      }
+      promise
+    } else {
+      f
+    }
+  }
+
+  private def dispatch(retryContext: RetryContext,
+                       request: Request,
+                       service: Service[Request, Response]): Future[Response] = {
+    val rep = service(request)
+    rep.transform { x =>
+      val classifier = x match {
+        case Throw(e) =>
+          retryContext.errorClassifier(e)
+        case Return(r) =>
+          retryContext.resultClassifier(r)
+      }
+
+      classifier match {
+        case ResultClass.Succeeded =>
+          rep
+        case ResultClass.Failed(isRetryable, cause) => {
+          if (retryContext.canContinue && isRetryable) {
+            Future
+              .value {
+                // Update the retry count
+                retryContext.nextRetry(cause)
+              }.flatMap { nextRetryContext =>
+                // Wait until the next retry
+                schedule(retryContext.nextWaitMillis.millis) {
+                  // Run the same request again
+                  dispatch(nextRetryContext, request, service)
+                }
+              }
+          } else {
+            // No more retry.
+            x match {
+              case Throw(e) =>
+                warn(e)
+                // Create an error response
+                val r = Response(Status.BadRequest)
+                r.setContentString(cause.getMessage)
```

Comment:
Currently, it looks that there is no way to handle this error message because `HttpClient` has only methods to treat response as JSON. Probably a decoding error happens in this case?

Reply:
That's true. This error response will be decoded as a regular response and will through another exception. Let me think about what is a good manner to handle such errors.

Reply:
Changed the logic to throw an exception 2b52331